visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp

```diff
@@ -0,0 +1,671 @@
+                         unsigned int max_iter )
+{
+  typedef typename RealPoint::Coordinate Scalar;
+  RealPoint gx;
+  Scalar f, g2;
+  Scalar eps2 = epsilon * epsilon;
+  while ( max_iter-- != 0 )
+    {
+      f = is( x );
+      if ( abs( f ) < epsilon ) return x;
+      gx = is.gradient( x );
+      g2 = gx.dot( gx );
+      if ( g2 < eps2 ) return x;
+      x -= (f/g2) * gx;
+    }
+  return x;
+}
+
+
+template <typename CubicalComplex4, typename ImplicitShape4, 
+          typename ImplicitDigitalShape4, typename ImplicitShape3>
+void projectComplex( std::vector< typename ImplicitShape3::RealPoint >& points,
+                     const CubicalComplex4& complex4,
+                     const ImplicitShape4& shape,
+                     const ImplicitDigitalShape4& dshape,
+                     const ImplicitShape3& shape3,
+                     double epsilon,
+                     unsigned int max_iter )
+{
+  typedef typename CubicalComplex4::Cell     Cell4;
+  typedef typename CubicalComplex4::Point    Point4;
+  typedef typename CubicalComplex4::CellMapConstIterator CellMapConstIterator;
+  typedef typename ImplicitShape4::RealPoint RealPoint4;
+  typedef typename ImplicitShape4::Ring      Ring;
+  typedef typename ImplicitShape3::RealPoint RealPoint3;
+  points.clear();
+  for ( CellMapConstIterator it = complex4.begin( 0 ), itE = complex4.end( 0 ); it != itE; ++it )
+    {
+      Cell4 cell    = it->first;
+      Point4 dp     = complex4.space().uCoords( cell );
+      RealPoint4 p  = dshape->embed( dp );
+      RealPoint3 p3 = RealPoint3( p[ 0 ], p[ 1 ], p[ 2 ] );
+      RealPoint3 q  = projectNewton( shape3, p3, epsilon, max_iter );
+      points.push_back( q );
+    }
+}
+
+template <typename CubicalComplex4, typename ImplicitShape4, 
+          typename ImplicitDigitalShape4, typename ImplicitShape3>
+void doNotProjectComplex( std::vector< typename ImplicitShape3::RealPoint >& points,
+                          const CubicalComplex4& complex4,
+                          const ImplicitShape4& shape,
+                          const ImplicitDigitalShape4& dshape,
+                          const ImplicitShape3& shape3 )
+{
+  typedef typename CubicalComplex4::Cell     Cell4;
+  typedef typename CubicalComplex4::Point    Point4;
+  typedef typename CubicalComplex4::CellMapConstIterator CellMapConstIterator;
+  typedef typename ImplicitShape4::RealPoint RealPoint4;
+  typedef typename ImplicitShape4::Ring      Ring;
+  typedef typename ImplicitShape3::RealPoint RealPoint3;
+  for ( CellMapConstIterator it = complex4.begin( 0 ), itE = complex4.end( 0 ); it != itE; ++it )
+    {
+      Cell4 cell    = it->first;
+      Point4 dp     = complex4.space().uCoords( cell );
+      RealPoint4 p  = dshape->embed( dp );
+      RealPoint3 p3 = RealPoint3( p[ 0 ], p[ 1 ], p[ 2 ] );
+      points.push_back( p3 );
+    }
+}
+    
+
+int main( int argc, char** argv )
+{
+  typedef int                               Integer;
```

Comment:
Here the problem is that I use both 3D and 4D objects, hence all these typedefs, which is quite heavy I agree.